### DISCUSSION TITLE
Make target-typed new objects callable

### DISCUSSION THREAD

**Discussion by fandrei** [dotnet/csharplang#8890] [Language Ideas]
At the current moment, it's possible to write the code like this:
```C#
static readonly DateTime Test1 = (new DateTime(2020, 01, 01)).ToUniversalTime();
```
However, not like this:
```C#
static readonly DateTime Test2 = (new(2020, 01, 01)).ToUniversalTime();
```
I think that it would be useful to make it work in the second case, too.

**Comment by BlinD-HuNTeR** [comment]
I like this thing, but the problem is, you could only call methods that would return the exact same type of the variable.

**Reply by jnm2** [reply]
Yeah, and it also raises questions like: how long of a chain of calls could you construct? And how do you deal with overload resolution between methods that might or might not return the exact same type, depending on the exact types of the arguments passed?

**Reply by fandrei** [reply]
> Yeah, and it also raises questions like: how long of a chain of calls could you construct? And how do you deal with overload resolution between methods that might or might not return the exact same type, depending on the exact types of the arguments passed?

Is it much different from the case when type name is specified explicitly? :shrug:

**Reply by TahirAhmadov** [reply]
@fandrei the problem is, how do you determine the type? For example:
```cs
static readonly int Test1 = (new(2020, 01, 01)).Year;
static readonly DateTime Test2 = new(2020, 01, 01) + TimeSpan.FromDays(1);
static readonly TimeSpan Test3 = new(2020, 01, 01) - new(2010, 01, 01);
```
Should the above scenarios be supported? If so, how would you deduce that it should be the `DateTime` constructor in the `new(...)`?

**Reply by fandrei** [reply]
> the problem is, how do you determine the type?

@TahirAhmadov we can just assume that `new()` with no type means the type specified on the left side of the assignment. After all, that's what target-typed new is supposed to mean.

**Comment by TahirAhmadov** [comment]
There definitely was an item requesting exactly this, and it was concluded that the complications here far outweigh the benefits. I can't find it, though.

**Reply by jnm2** [reply]
Yes, https://github.com/dotnet/csharplang/discussions/5211 and https://github.com/dotnet/csharplang/discussions/6942.
Similar type of request: https://github.com/dotnet/csharplang/discussions/4959 and https://github.com/dotnet/csharplang/discussions/5561

**Reply by fandrei** [reply]
@jnm2 just assume that `new()` means the same type as the type of the assignment target. I don't see problems with such approach.

**Reply by jnm2** [reply]
Sure :)

**Reply by fandrei** [reply]
> Sure :)

Do you see such problems? :shrug:

**Reply by CyrusNajmabadi** [reply]
I personally don't think it makes much sense.   You have a tree like so:

```
                x
             /     \
           y         z
         /    \  
    new()      etc
```

I don't see why this node shoudl take on meaning well outside the tree to define what it means.

**Reply by fandrei** [reply]
> I don't see why this node shoudl take on meaning well outside the tree to define what it means.

Isn't this the same thing that target-typed new does?

**Reply by HaloFour** [reply]
There are lots of things that the compiler could do, the question is whether or not it should and whether or not it's worth the effort to do so.  IMO, allowing arbitrary dotted expressions off of a `new` operator decreases clarity.  The compiler could definitely infer what it's supposed to do unambiguously, but I'm much more concerned about the human parser.

**Reply by fandrei** [reply]
> but I'm much more concerned about the human parser

Oh yeah. What does `new(new(var1))` mean?

**Reply by CyrusNajmabadi** [reply]
> Isn't this the same thing that target-typed new does?

No.  it's not a random subnode that is targeted.  It is the new() itself.  It's always hte top node.

**Reply by CyrusNajmabadi** [reply]
> Oh yeah. What does new(new(var1)) mean?

The outer 'new' is target typed to determine it's type.  This then becomes a call to `new WhateverType(new(...))`  Then, overload resolution goes to find a proper constructor, and uses that to target type the inner `new(...)`

**Reply by HaloFour** [reply]
> Oh yeah. What does new(new(var1)) mean?

Indeed, I would argue that it's pretty awful, but that falls out of allowing for target-typed `new()` in general.  Something that is pretty bad doesn't justify doing more that is pretty bad.

**Reply by fandrei** [reply]
> The outer 'new' is target typed to determine it's type. This then becomes a call to `new WhateverType(new(...))` Then, overload resolution goes to find a proper constructor, and uses that to target type the inner `new(...)`

I know how it works. However, human parsers (mentioned above) aren't very good with this construct. Especially if you add more arguments and function overloading. In some examples of the code that I've seen, it was confusing af. However, this probably deserves a separate discussion.

**Reply by fandrei** [reply]
> Something that is pretty bad doesn't justify doing more that is pretty bad.

It's not bad. You can easily see what the type in the feature that I proposed would be.

**Reply by fandrei** [reply]
> It is the new() itself. It's always hte top node.

Ok. So, I don't see any problems if it would be possible to call methods in this case and not in other ones.

**Reply by CyrusNajmabadi** [reply]
> I know how it works.

You asked what it meant.  I was explaining :)

>However, human parsers (mentioned above) aren't very good with this construct.

I don't disagree with that.  That's why we do not recommend changing your code to look like that in the IDE.

> It's not bad. You can easily see what the type in the feature that I proposed would be.

It's unclear to me why the type of the left child of some member access (or any other construct) would be inferred this way.

**Reply by CyrusNajmabadi** [reply]
> Ok. So, I don't see any problems if it would be possible to call methods in this case and not in other ones.

I didn't say it was a problem. I said i didn't think it made sense from a language perspective.  If i have some arbitrary tree of nodes, having one leftmost child take on this meaning isn't something i think is sensible.  

In order to adopt a rule, i want to feel that the rule makes sense and is a good thing.

**Reply by jnm2** [reply]
> Do you see such problems? 🤷

@fandrei Our first conversation going in circles has set an expectation. I'm not really attracted by the setup of "here's an idea I'm attached to, and it's going to be strenuous to change my mind." That's work that you can do on your own, without me slowing down your process :)

**Reply by fandrei** [reply]
> conversation going in circles

Nope. However, I'm not a fan of made up arguments why some feature is not necessary (made up arguments why some feature is necessary, too).
I'm trying to figure out why the C# team guys consider some features unnecessary (even though they would save some boilerplate code, and were asked for by a number of people), while they consider other features necessary (even though they don't solve any problems at all).

**Reply by jnm2** [reply]
I've neither claimed that a feature is unnecessary nor that any feature is necessary.

**Reply by fandrei** [reply]
> It's unclear to me why the type of the left child of some member access (or any other construct) would be inferred this way.

Just pretend that `DateTime Test1 = (new(2020, 01, 01)).ToUniversalTime();` in fact means `DateTime Test1 = (new DateTime(2020, 01, 01)).ToUniversalTime();`. You already know the type of this new.

**Reply by CyrusNajmabadi** [reply]
> I'm trying to figure out why the C# team guys 

I've given you my stated reasons.  You are choosing to reject them based on your own views.  That's your call. 

> while they consider other features necessary (even though they don't solve any problems at all).

That may be your perspective.  That is not ours.

**Reply by CyrusNajmabadi** [reply]
> Just pretend

As mentioned, i don't have any feeling that thsi should be the pretended interpretation of this syntax.  And, as mentioned elsewhere (in our discord), i actually think that syntax should mean something else *entirely*.  As such, i'm not in favor of this idea.  Furthermore, i don't see this properly composing, as i would expect from normal lang features.  for example, this would not work:

`DateTime test1 = x ? new(2020, 01, 01)).ToUniversalTime() : new(2021, 01, 01)).ToUniversalTime();`

That makes it a very odd feature, and one i would be opposed to.

**Reply by TahirAhmadov** [reply]
@fandrei the problem is, if we were to aim to support all possible permutations (like in my comment above with a few samples), it runs into the issue of "how do we determine the type". However, if we were to limit it to just "same type on the left and right", it becomes a very complex exercise of coming up with rules of what is and isn't allowed; it becomes an expression of a different kind than any other expression in C#. 
All of it just to save one keystroke (Tab or Enter to accept Intellisense suggestion) after typing `new `. 😕

**Reply by fandrei** [reply]
>  You are choosing to reject them based on your own views.

Nope, based on facts.

> DateTime test1 = x ? new(2020, 01, 01)).ToUniversalTime() : new(2021, 01, 01)).ToUniversalTime();

So, let's get back to the facts.
We know the type in this case too, right? Let's just interpret `new()` as `(new KnownType())`. Do you see any problems with such interpretation?

**Reply by CyrusNajmabadi** [reply]
> Nope, based on facts.

This conversation isn't going to be productive if you keep acting in this fashion.  You are asking for our subjective views, and preferring your own.  Rejecting these perspectives because you deem your views as the only facts that are acceptable it not discourse.

You said:

> I'm trying to figure out why the C# team guys

If you are not willing to actually listen when we give you those views, then there is no point in continuing.  If you express a willingness to actually consider these perspectives, let me know.

**Reply by fandrei** [reply]
> You are asking for our subjective views

I never asked for subjective views.

> This conversation isn't going to be productive

Ok, let's keep it productive. Do you see any problems with this approach?

> Let's just interpret new() as (new KnownType()). Do you see any problems with such interpretation?

**Reply by CyrusNajmabadi** [reply]
> I never asked for subjective views.

You literally said:

>I'm trying to figure out why the C# team guys consider some features unnecessary

This is a question on how we subjectively rank thinks.  If you ask that question, you will get our subjective views on what makes one feature less necessary than another.

**Reply by HaloFour** [reply]
https://github.com/dotnet/csharplang/discussions/8890#discussioncomment-11608974

**Reply by fandrei** [reply]
> This is a question on how we subjectively rank thinks.

No, I'm mostly interested in arguments "for" and "against", and comparison of such arguments.
However, if your arguments look weak to me - I will write about that and explain why.
I don't have to just accept your arguments as is, right?

**Reply by CyrusNajmabadi** [reply]
> No, I'm mostly interested in arguments "for" and "against", and comparison of such arguments.

This is subjective.   If you ask for subjective responses, it's not ok to then reject them for being subjective.

>However, if your arguments look weak to me

That's not my concern.  You asked for opinions.  Opinions are being provided.  I personally have no interest in trying to convince you that my personal opinion is correct.

**Reply by fandrei** [reply]
> Opinions are being provided. I personally have no interest in trying to convince you that my personal opinion is correct.

So, I have to agree with your opinion and cannot question the arguments that you supported it with. Got it.
